### PR TITLE
Add README with proposed target for what this could look like.

### PR DESCRIPTION
**Pull request by dgoodwin**
Please mark up with review as necessary.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      yedit:
+        key: spec.metadata.name
+        value: "{{ replacement_name }}"
+        value_type: list
+    - src: patchme.yml
+      # Support supplying structured patches with "kubectl patch" prior to submitting to the server.
+      patches:
+      - patch1.yml
+      - patch2.yml
+    # This would be nice too, kubectl supports, would likely imply no yediting or patching unless
```
Comment (dgoodwin) [inline README.md:82]: Assumption here is that if you want to yedit or patch, you need to nest that with a specific file.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    # we want the module to be in the business of fetching files.
+    - src: https://github.com/me/project/template.yml
+
+    # Support templating files out automatically. Use Ansible's native mechanisms for jinja
+    # templates but don't require the user to copy the file and manually clean it up with
+    # their own tasks.
+    templates:
+    - src: apptemplate.yml
+      vars:
+        foo: bar
```
Comment (dgoodwin) [inline README.md:92]: In theory we could allow yedit/patch on the templates as well but may not be necessary.

Comment (dgoodwin) [reply README.md:92]: I am curious if Ansible core is open to a module with templating embedded, or if this is not acceptable and users would be expected to manually use the template module, and cleanup files themselves. As a user, I'd love to see this handled for me.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    - src: subdir/
+    - src: configmap.json
+    inline: |
+      kind: "Namespace"
+      apiVersion: v1
+      metadata:
+        name: testnamespace
+
+
+- kubectl_facts:
```
Comment (smarterclayton) [inline README.md:119]: For this, any reason to bias towards `kubectl` vs `kubernetes_facts`?  The implementation is somewhat arbitrary (in the future could be python).

Everything you described below is almost a direct server API call, so some flexibility there.

Also, never use `kind` for `configmap`.  A kind is `ConfigMap` which is a schema.  A `configmap` would be a resource.  This should be using a `resource` string.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    - src: apptemplate.yml
+      vars:
+        foo: bar
+
+    # Control what binary to use, as oc is required for apply operations on OpenShift types:
+    binary: oc
+
+    # Various other kubectl options could be supported:
+    prune: true
+    dry_run: true
```
Comment (smarterclayton) [inline README.md:99]: Would this be a `state`?  What's the ansible equiv to dry run?  `check`?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    # Support templating files out automatically. Use Ansible's native mechanisms for jinja
+    # templates but don't require the user to copy the file and manually clean it up with
+    # their own tasks.
+    templates:
+    - src: apptemplate.yml
+      vars:
+        foo: bar
+
+    # Control what binary to use, as oc is required for apply operations on OpenShift types:
+    binary: oc
```
Comment (smarterclayton) [inline README.md:95]: Hrm, in theory we don't have to do this today, but starting soon we might.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    # then submitted with apply.
+    # Specifying multiple files implies one execution of kubectl apply. If you want multiple
+    # you can use separate tasks or a with_items loop.
+    files:
+    - src: configmap.json
+    - src: subdir/
+    - src: service.yml
+      # Support small edits to the yaml on the fly, without going to a full template. Operates
+      # on the temporary copy of the file sent to the remote host.
+      yedit:
```
Comment (smarterclayton) [inline README.md]: I think it's probably better to call out yedit as a separate, valuable module.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+languages. In the future this logic may move to the API, however this is not
+expected near term.
+
+However kubectl is intended to be treated as a type of API, as such in the interim we can leverage it's logic by wrapping in Ansible modules.
+
+This prototype is specific to generically managing actual Kubernetes config, it
+does not wrap Kubernetes yaml in Ansible yaml, or expose individual modules per
+individual Kubernetes type.
+
+This prototype aims to make applying Kubernetes config via Ansible extremely
```
Comment (smarterclayton) [inline README.md]: Would add "consistent with the expected experience for config management in kuberentes" and link to some of brian's docs

File: README.md (hunk 3 above)
Comment (dgoodwin) [reply README.md:119]: That's interesting. The thought was to clearly identify that this is talking to kubectl on the target host, not talking to the actual Kube REST API from the target host. Could we just flip users from one to the other in the future? It seems like it *might* be theoretically possible but I'm not sure it's the kind of thing we could do without upsetting people, although in essence it's not much different, still the same host talking to the same REST API at the end of the day.

Comment (dgoodwin) [reply README.md:119]: I was assuming 'kind' was case insensitive as we do in some of these pre-existing modules or on CLI, are you commenting specifically that it should be case sensitive, or that "kind" literally should not be an option here, i.e. should it be:

```
kind: ConfigMap
resource: myconfig (assuming you want a specific resource not a list)
```

File: README.md (hunk 5 above)
Comment (dgoodwin) [reply README.md:95]: Oh I thought you had mentioned to get at the apply logic for some of our types, we'd need to switch to oc binary.

File: README.md (hunk 4 above)
Comment (dgoodwin) [reply README.md:99]: Ansible uses --check yes, we could make the module identify as supporting check and if try do a kubectl apply dry-run, assuming we get apply returning some indication if something changed. I will update.

File: README.md (hunk 6 above)
Comment (dgoodwin) [reply README.md]: Could it still be embedded for use here just like templating and patching? If not, users now need to think about writing files out to specific locations on the remote host and cleaning them up, whereas I'd like to see this just handle all of that for you. User likely doesn't care about having to place files on remote server, then apply them in separate ansible tasks, then potentially clean them up.

**Comment by dgoodwin** [comment]
Some updates based on feedback, still not a final form but merging as a first draft.

**Pull request merged**